src/tools/compiletest/src/runtest.rs

```diff
@@ -1885,34 +1862,47 @@ impl<'test> TestCx<'test> {
             .unwrap();
     }
 
+    /// Create a filename for output with the given extension.  Example:
+    ///   /.../testname.revision.mode/testname.extension
     fn make_out_name(&self, extension: &str) -> PathBuf {
         self.output_base_name().with_extension(extension)
     }
 
+    /// Directory where auxiliary files are written.  Example:
+    ///   /.../testname.revision.mode/auxiliary/
     fn aux_output_dir_name(&self) -> PathBuf {
-        self.output_base_name_stage()
+        self.output_base_dir()
+            .join("auxiliary")
             .with_extra_extension(self.config.mode.disambiguator())
-            .with_extra_extension(".aux")
     }
 
-    fn output_testname(&self, filepath: &Path) -> PathBuf {
-        PathBuf::from(filepath.file_stem().unwrap())
+    /// Generates a unique name for the test, such as `testname.revision.mode`.
+    fn output_testname_unique(&self) -> PathBuf {
+        output_testname_unique(self.config, self.testpaths, self.safe_revision())
     }
 
-    /// Given a test path like `compile-fail/foo/bar.rs` returns a name like
-    /// `/path/to/build/<triple>/test/compile-fail/foo/bar`.
-    fn output_base_name(&self) -> PathBuf {
-        let dir = self.config.build_base.join(&self.testpaths.relative_dir);
+    /// The revision, ignored for Incremental since it wants all revisions in
+    /// the same directory.
```

Comment:
All these special cases definitely make me think that separating out the incremental *phases* from *revisions* would be a good idea -- but maybe in a follow-up PR.